pkg/products/grafana/reconciler.go

```diff
@@ -0,0 +1,320 @@
+	productNamespace := r.Config.GetNamespace()
+
+	phase, err := r.ReconcileFinalizer(ctx, client, installation, string(r.Config.GetProductName()), func() (integreatlyv1alpha1.StatusPhase, error) {
+		phase, err := resources.RemoveNamespace(ctx, installation, client, productNamespace)
+		if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+			return phase, err
+		}
+
+		phase, err = resources.RemoveNamespace(ctx, installation, client, operatorNamespace)
+		if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+			return phase, err
+		}
+
+		return integreatlyv1alpha1.PhaseCompleted, nil
+	})
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, "Failed to reconcile finalizer", err)
+		return phase, err
+	}
+
+	phase, err = r.ReconcileNamespace(ctx, operatorNamespace, installation, client)
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, fmt.Sprintf("Failed to reconcile %s ns", operatorNamespace), err)
+		return phase, err
+	}
+
+	err = r.createSecrets(ctx, client, installation, &grafanav1alpha1.Grafana{})
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, fmt.Sprintf("Failed to reconcile %s ns", productNamespace), err)
+		return phase, err
+	}
+
+	phase, err = r.reconcileSubscription(ctx, client, installation, productNamespace, operatorNamespace)
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, fmt.Sprintf("Failed to reconcile %s subscription", constants.ThreeScaleSubscriptionName), err)
+		return phase, err
+	}
+
+	phase, err = r.reconcileComponents(ctx, client, installation, &grafanav1alpha1.Grafana{})
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, "Failed to create components", err)
+		return phase, err
+	}
+
+	return integreatlyv1alpha1.PhaseCompleted, nil
+}
+func (r *Reconciler) createSecrets(ctx context.Context, client k8sclient.Client, installation *integreatlyv1alpha1.RHMI, cr *grafanav1alpha1.Grafana) error {
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "grafana-k8s-proxy",
+			Namespace: r.Config.GetOperatorNamespace(),
+		},
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, client, secret, func() error {
+		owner.AddIntegreatlyOwnerAnnotations(secret, installation)
+		if secret.Data == nil {
+			secret.Data = map[string][]byte{}
+		}
+		secret.Data["session_secret"] = []byte(PopulateSessionProxySecret())
+		return nil
+	})
+
+	if err != nil {
+		logrus.Errorf("failed to create grafana proxy secret, %s: %w", secret, err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) reconcileComponents(ctx context.Context, client k8sclient.Client, installation *integreatlyv1alpha1.RHMI, cr *grafanav1alpha1.Grafana) (integreatlyv1alpha1.StatusPhase, error) {
```

Comment:
Is it required to pass the cr here? I don't see it used in the function.

Reply:
Good spot, Have removed it.